railties/lib/rails/test_unit/runner.rb

```diff
@@ -60,8 +60,10 @@ def compose_filter(runnable, filter)
 
         private
           def extract_filters(argv)
-            # Extract absolute and relative paths but skip -n /.*/ regexp filters.
-            argv.select { |arg| %r%^/?\w+/%.match?(arg) && !arg.end_with?("/") }.map do |path|
+            argv.select do |arg|
+              # Extract absolute and relative paths but skip -n /.*/ regexp filters.
+              %r%^/?\w+/%.match?(arg) && !(arg.start_with?("/") && arg.end_with?("/"))
```

Comment:
Wanted to try an alternative with a negative lookahead, `(?!=$)`. The idea being that we a regexp filter in case we haven't encountered a previous / and have hit reached the end of the string:

```
>> /^\/?\w+\/(?!=$)/.match? "/path/to/test"
=> true
>> /^\/?\w+\/(?!=$)/.match? "/path/to/test/"
=> true
>> /^\/?\w+\/(?!=$)/.match? "/path to test/"
=> false
>> /^\/?\w+\/(?!=$)/.match? "/path to test"
=> false
>> /^\/?\w+\/(?!=$)/.match? "/path/to/test"
=> true
>> /^\/?\w+\/(?!=$)/.match? "path/to/test"
=> true
>> /^\/?\w+\/(?!=$)/.match? "path/to/test/"
=> true
```

It's still confusing and I'm wondering if we could extract a `relative_or_absolute_path_without_matching_regexp_filter?` helper or a `regexp_filter?` that we `reject` instead.

Reply:
Another idea could be to try something like `arg.scan(/\w+\/(?!=$)/).any?`. Though I'm a little tired right now, so maybe someone brighter will know? 😂

Reply:
I believe negative lookahead can do the job:

```ruby
["/ok", "ok/", "ok/ok", "/ok/ok/", "/no/", "no"].grep %r"^(?!/[^/]+/$).*/"
# == ["/ok", "ok/", "ok/ok", "/ok/ok/"]
```

Reply:
We now have three proposed solutions with three different behaviours. 🙂 

```ruby
irb(main):001:0> ["/ok", "ok/", "ok/ok", "/ok/ok/", "/no/", "no"].select { |arg| %r%^/?\w+/%.match?(arg) && !arg.end_with?("/") } # current behaviour
=> ["ok/ok"]
irb(main):002:0> ["/ok", "ok/", "ok/ok", "/ok/ok/", "/no/", "no"].select { |arg| %r%^/?\w+/%.match?(arg) && !(arg.start_with?("/") && arg.end_with?("/")) } # my solution
=> ["ok/", "ok/ok"]
irb(main):003:0> ["/ok", "ok/", "ok/ok", "/ok/ok/", "/no/", "no"].grep /^\/?\w+\/(?!=$)/ # Kasper's solution
=> ["ok/", "ok/ok", "/ok/ok/", "/no/"]
irb(main):004:0> ["/ok", "ok/", "ok/ok", "/ok/ok/", "/no/", "no"].grep %r"^(?!/[^/]+/$).*/" # Jonathan's solution
=> ["/ok", "ok/", "ok/ok", "/ok/ok/"]
```

I'm not sure we should be treating `/ok/ok/` as a path; regex filters can include slashes:

```bash
$ cd activemodel; bin/test test/cases/attribute_methods_test.rb -n /private/protected/
Run options: -n /private/protected/ --seed 15071

# Running:

..

Finished in 0.001028s, 1945.5253 runs/s, 7782.1014 assertions/s.
2 runs, 8 assertions, 0 failures, 0 errors, 0 skips
```

I extracted the conditions to two private methods; interested to hear what you both think.

Reply:
> We now have three proposed solutions with three different behaviours. 🙂

I think we can get it to 10 if we really push ourselves 😂

Reply:
> I think we can get it to 10 if we really push ourselves

**`CHALLENGE ACCEPTED.`**

If we're supporting absolute paths, I think treating `/ok/ok/` as a path would be preferable.  While test names can contain slashes, I feel they are not so common.  And needing to filter precisely on the slashes themselves seems less common still.  For example, `/private/protected/` could be written as `/private.protected/`.

However, if we decide to treat `/ok/ok/` as a pattern, we could still use a Regexp with negative lookahead:

```ruby
args = ["/ok", "ok/", "ok/ok", "/ok/ok", "ok/ok/", "/maybe/maybe/", "/no/", "no"]

args.grep %r"^(?!/[^/]+/$).*/"
# == ["/ok", "ok/", "ok/ok", "/ok/ok", "ok/ok/", "/maybe/maybe/"]

args.grep %r"^(?!/.+/$).*/"
# == ["/ok", "ok/", "ok/ok", "/ok/ok", "ok/ok/"]
```

Reply:
You're probably right that passing an absolute path with a trailing slash is more likely than using a regex pattern that contains a slash. I still think we should be conservative here, since Rails' test runner is the interloper, manually parsing arguments that are intended for Minitest.

There may be a way to support both cases: Minitest calls `OptionParser#parse!`, which removes options from the input array as it parses them:

https://github.com/seattlerb/minitest/blob/6257210b7accfeb218b4388aaa36d3d45c5c41a5/lib/minitest.rb#L213

This means that by the time Minitest calls our plugin's init hook, ARGV will only contain positional arguments, and we could confidently treat all of them as paths:

https://github.com/seattlerb/minitest/blob/6257210b7accfeb218b4388aaa36d3d45c5c41a5/lib/minitest.rb#L128
https://github.com/seattlerb/minitest/blob/6257210b7accfeb218b4388aaa36d3d45c5c41a5/lib/minitest.rb#L135
https://github.com/rails/rails/blob/f3dfed7d0634bb7c07d761c62c11065d82488941/railties/lib/minitest/rails_plugin.rb#L40

This behaviour isn't documented by Minitest, though. There was a proposal to change it in https://github.com/seattlerb/minitest/pull/369, and it looks like it only died because the author lost interest. I may ask if Minitest can expose the leftover arguments to plugins so that we can depend on them.

Anyway, given that this patch fixes a problem I actually ran into in practice, I'm going to merge it. If someone complains about the absolute-path-with-trailing-slash case, we can reconsider.